test/fiber/scheduler.rb

```diff
@@ -136,13 +135,51 @@ def wait_any(io, events, duration)
   end
 
   def wait_for_single_fd(fd, events, duration)
-    wait_any(
-      for_fd(fd),
-      events,
-      duration
-    )
+    wait_any(for_fd(fd), events, duration)
   end
 
+  # called from Mutex#lock
+  def wait_mutex(mutex)
+    p [:wait_mutex, mutex]
+    # @waiting_on[Fiber.current] = mutex
+
+    # Guarantee: Fiber.current is a Fiber of current Thread and is a Fiber of this scheduler
+    @waiting_mutex[mutex] << Fiber.current
+    Fiber.yield
+    true
+  end
+
+  # called from Mutex#unlock, possibly from another Fiber on the same thread, or possibly from some other thread
+  def notify_mutex(mutex)
+    p [:notify_mutex, mutex]
+    # p @waiting_mutex
+    q = @waiting_mutex[mutex]
+    fiber = q.shift
+    @waiting_mutex.delete(mutex) if q.empty?
+    raise unless fiber
+
+    if Thread.current == @thread
+      # but what if fiber died in the meanwhile?
+      # Is it possible if same thread? Yes, due to Fiber#raise
+      # And what if it somehow went further in the meanwhile, due to e.g. Fiber#raise, we'll resume another point?
+      if fiber.alive?
+        p [:resume, fiber]
+        fiber.resume
+      end
+    else
+      # scheduler = Thread.current.scheduler
+      # send -> { notify_mutex(mutex) } to scheduler Queue + control pipe write to wake up?
+      # (could be abstracted as #external_notification(&block) ?)
+      # or maybe shouldn't come at all in #notify_mutex ? => seems simpler
+      # Best seems to remember scheduler+Mutex instance (+Fiber?) in `waitq` for non-blocking waiters,
+      # so we know exactly how to wake up a given waiter Fiber.
+      raise "different thread"
+    end
+  end
+
+  # BTW, super unsafe if e.g. some IO like DB connection is shared between Fibers, no?
+  # (e.g., send, send, recv, recv, might be reordered)
```

Comment:
This is a more general question, not really related to this PR.
I guess it's tricky to reuse existing thread pools such as in Sequel, and could be dangerous with the scheduler. Probably need per-Fiber connection, and then some way to limit max number of concurrent connections?
It seems unfortunate that existing code might trigger quite dangerous behavior (e.g., getting the response for another DB query) with the scheduler.

Reply:
I think this is something where libraries will have to adapt.

Libraries that assume a Thread is the smallest context of execution are wrong. It's Fiber and has been for a long time.

Libraries that pass around implicit state, e.g. ActiveRecord, are fundamentally broken in this regard. I'm less concerned about this issue since it's a design choice of the library. The introduction of the scheduler doesn't cause this problem, it just exposes a fundamentally wrong design decision in a more obvious way.